From 3e83a55303ecda4fa79d72fdf14988d4b4235404 Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Mon, 1 Nov 2021 09:00:29 +0100
Subject: [PATCH] nano: Integrate with Balena and load kernel from rootfs

This commit:
1) Integrates balenaOS with the Jetson Nano platform
2) Adapts to L4T 32.6.1 u-boot 2020.04 from upstream Nvidia.
   We previously switched to this from upstream meta-tegra because this
   revision works better with custom device trees.
3) We add a separate config for Jetson Nano eMMC, like
   upstream meta-tegra does in
   https://github.com/OE4T/u-boot-tegra/commit/31554c1e8f940d2340065af8d565525096fa0957

   This helps us diferentiate the Nano SD from Nano eMMC
   when generating the boot blob (stored in qspi-nor or on hw part)

Upstream-status: Inappropriate [configuration]
Signed-off-by: Alexandru Costache <alexandru@balena.io>
---
 board/nvidia/p3450-0000/Kconfig |  3 ++
 configs/p3450-0000_defconfig    | 66 +++++++++++++++------------
 configs/p3450-0002_defconfig    | 80 +++++++++++++++++++++++++++++++++
 configs/p3541-0000_defconfig    | 63 ++++++++++++++------------
 include/config_distro_bootcmd.h | 14 ++++--
 include/configs/p3450-0000.h    | 25 +++++++----
 include/configs/p3541-0000.h    |  4 +-
 7 files changed, 184 insertions(+), 71 deletions(-)
 create mode 100644 configs/p3450-0002_defconfig

diff --git a/board/nvidia/p3450-0000/Kconfig b/board/nvidia/p3450-0000/Kconfig
index 7a08cd8867..033718d232 100644
--- a/board/nvidia/p3450-0000/Kconfig
+++ b/board/nvidia/p3450-0000/Kconfig
@@ -9,4 +9,7 @@ config SYS_VENDOR
 config SYS_CONFIG_NAME
 	default "p3450-0000"
 
+config P3450_EMMC
+	bool "SKU 2 module with eMMC"
+
 endif
diff --git a/configs/p3450-0000_defconfig b/configs/p3450-0000_defconfig
index 9e84984135..c9d5a606e9 100644
--- a/configs/p3450-0000_defconfig
+++ b/configs/p3450-0000_defconfig
@@ -15,14 +15,14 @@ CONFIG_CMD_DFU=y
 CONFIG_CMD_GPIO=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_MMC=y
-CONFIG_CMD_PCI=y
+#CONFIG_CMD_PCI=y
 CONFIG_CMD_SF=y
 CONFIG_CMD_SPI=y
-CONFIG_CMD_USB=y
-CONFIG_CMD_USB_MASS_STORAGE=y
-# CONFIG_CMD_SETEXPR is not set
+#CONFIG_CMD_USB=y
+#CONFIG_CMD_USB_MASS_STORAGE=y
+CONFIG_CMD_SETEXPR=y
 # CONFIG_CMD_NFS is not set
-CONFIG_CMD_EXT4_WRITE=y
+#CONFIG_CMD_EXT4_WRITE=y
 CONFIG_OF_LIVE=y
 CONFIG_DEFAULT_DEVICE_TREE="tegra210-p3450-0000"
 CONFIG_DFU_MMC=y
@@ -34,38 +34,46 @@ CONFIG_SPI_FLASH_MACRONIX=y
 CONFIG_SPI_FLASH_USE_4K_SECTORS=y
 CONFIG_SF_DEFAULT_MODE=0
 CONFIG_SF_DEFAULT_SPEED=24000000
-CONFIG_RTL8169=y
-CONFIG_PCI=y
-CONFIG_DM_PCI=y
-CONFIG_DM_PCI_COMPAT=y
-CONFIG_PCI_TEGRA=y
+#CONFIG_RTL8169=y
+#CONFIG_PCI=y
+#CONFIG_DM_PCI=y
+#CONFIG_DM_PCI_COMPAT=y
+#CONFIG_PCI_TEGRA=y
 CONFIG_SYS_NS16550=y
 CONFIG_TEGRA114_SPI=y
 CONFIG_TEGRA210_QSPI=y
-CONFIG_USB=y
-CONFIG_DM_USB=y
-CONFIG_USB_EHCI_HCD=y
-CONFIG_USB_EHCI_TEGRA=y
-CONFIG_USB_GADGET=y
-CONFIG_USB_GADGET_MANUFACTURER="NVIDIA"
-CONFIG_USB_GADGET_VENDOR_NUM=0x0955
-CONFIG_USB_GADGET_PRODUCT_NUM=0x701a
+#CONFIG_USB=y
+#CONFIG_DM_USB=y
+#CONFIG_USB_EHCI_HCD=y
+#CONFIG_USB_EHCI_TEGRA=y
+#CONFIG_USB_GADGET=y
+#CONFIG_USB_GADGET_MANUFACTURER="NVIDIA"
+#CONFIG_USB_GADGET_VENDOR_NUM=0x0955
+#CONFIG_USB_GADGET_PRODUCT_NUM=0x701a
 CONFIG_CI_UDC=y
-CONFIG_USB_GADGET_DOWNLOAD=y
-CONFIG_USB_HOST_ETHER=y
-CONFIG_USB_ETHER_ASIX=y
-CONFIG_ENV_IS_IN_MMC=y
-CONFIG_ENV_IS_IN_SPI_FLASH=y
-CONFIG_ENV_SIZE=0x2000
+#CONFIG_USB_GADGET_DOWNLOAD=y
+#CONFIG_USB_HOST_ETHER=y
+#CONFIG_USB_ETHER_ASIX=y
+#CONFIG_ENV_IS_IN_MMC=y
+#CONFIG_ENV_IS_NOWHERE=y
+CONFIG_ENV_SIZE=0x4000
 CONFIG_ENV_SECT_SIZE=0x1000
 CONFIG_ENV_OFFSET=0xFFFD8000
-CONFIG_BOOTP_PREFER_SERVERIP=y
+#CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_POSITION_INDEPENDENT=y
 CONFIG_DISABLE_SDMMC1_EARLY=y
-CONFIG_NVME=y
-CONFIG_USB_XHCI_HCD=y
-CONFIG_USB_XHCI_TEGRA=y
-CONFIG_USB_STORAGE=y
+#CONFIG_NVME=y
+#CONFIG_USB_XHCI_HCD=y
+#CONFIG_USB_XHCI_TEGRA=y
+#CONFIG_USB_STORAGE=y
 CONFIG_DOS_PARTITION=y
 CONFIG_CMD_CACHE=y
 CONFIG_OF_LIBFDT_OVERLAY=y
+CONFIG_CMD_PART=y
+CONFIG_PARTITION_UUIDS=y
+CONFIG_CMD_IMPORTENV=y
+CONFIG_BOOTDELAY=0
+CONFIG_CMD_EXPORTENV=y
+CONFIG_CMD_FS_UUID=y
+CONFIG_AUTOBOOT=y
+
diff --git a/configs/p3450-0002_defconfig b/configs/p3450-0002_defconfig
new file mode 100644
index 0000000000..bd172b7154
--- /dev/null
+++ b/configs/p3450-0002_defconfig
@@ -0,0 +1,80 @@
+CONFIG_ARM=y
+CONFIG_TEGRA=y
+CONFIG_SYS_TEXT_BASE=0x80080000
+CONFIG_TEGRA210=y
+CONFIG_TARGET_P3450_0000=y
+CONFIG_NR_DRAM_BANKS=3
+CONFIG_OF_SYSTEM_SETUP=y
+CONFIG_OF_BOARD_SETUP=y
+CONFIG_CONSOLE_MUX=y
+CONFIG_SYS_STDIO_DEREGISTER=y
+CONFIG_SYS_PROMPT="Tegra210 (P3450-0002) # "
+# CONFIG_CMD_IMI is not set
+CONFIG_CMD_DFU=y
+# CONFIG_CMD_FLASH is not set
+CONFIG_CMD_GPIO=y
+CONFIG_CMD_I2C=y
+CONFIG_CMD_MMC=y
+#CONFIG_CMD_PCI=y
+CONFIG_CMD_SF=y
+CONFIG_CMD_SPI=y
+#CONFIG_CMD_USB=y
+#CONFIG_CMD_USB_MASS_STORAGE=y
+CONFIG_CMD_SETEXPR=y
+# CONFIG_CMD_NFS is not set
+#CONFIG_CMD_EXT4_WRITE=y
+CONFIG_OF_LIVE=y
+CONFIG_DEFAULT_DEVICE_TREE="tegra210-p3450-0000"
+CONFIG_DFU_MMC=y
+CONFIG_DFU_RAM=y
+CONFIG_DFU_SF=y
+CONFIG_SYS_I2C_TEGRA=y
+CONFIG_SPI_FLASH=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_SPI_FLASH_USE_4K_SECTORS=y
+CONFIG_SF_DEFAULT_MODE=0
+CONFIG_SF_DEFAULT_SPEED=24000000
+#CONFIG_RTL8169=y
+#CONFIG_PCI=y
+#CONFIG_DM_PCI=y
+#CONFIG_DM_PCI_COMPAT=y
+#CONFIG_PCI_TEGRA=y
+CONFIG_SYS_NS16550=y
+CONFIG_TEGRA114_SPI=y
+CONFIG_TEGRA210_QSPI=y
+#CONFIG_USB=y
+#CONFIG_DM_USB=y
+#CONFIG_USB_EHCI_HCD=y
+#CONFIG_USB_EHCI_TEGRA=y
+#CONFIG_USB_GADGET=y
+#CONFIG_USB_GADGET_MANUFACTURER="NVIDIA"
+#CONFIG_USB_GADGET_VENDOR_NUM=0x0955
+#CONFIG_USB_GADGET_PRODUCT_NUM=0x701a
+CONFIG_CI_UDC=y
+#CONFIG_USB_GADGET_DOWNLOAD=y
+#CONFIG_USB_HOST_ETHER=y
+#CONFIG_USB_ETHER_ASIX=y
+CONFIG_ENV_IS_NOWHERE=y
+#CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_ENV_SIZE=0x4000
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_ENV_OFFSET=0xFFFD8000
+#CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_POSITION_INDEPENDENT=y
+CONFIG_DISABLE_SDMMC1_EARLY=y
+#CONFIG_NVME=y
+#CONFIG_USB_XHCI_HCD=y
+#CONFIG_USB_XHCI_TEGRA=y
+#CONFIG_USB_STORAGE=y
+CONFIG_DOS_PARTITION=y
+CONFIG_CMD_CACHE=y
+#CONFIG_USB_GADGET_DOWNLOAD=y
+#CONFIG_USB_HOST_ETHER=y
+CONFIG_P3450_EMMC=y
+CONFIG_CMD_PART=y
+CONFIG_PARTITION_UUIDS=y
+CONFIG_CMD_IMPORTENV=y
+CONFIG_BOOTDELAY=0
+CONFIG_CMD_EXPORTENV=y
+CONFIG_CMD_FS_UUID=y
+CONFIG_AUTOBOOT=y
diff --git a/configs/p3541-0000_defconfig b/configs/p3541-0000_defconfig
index 9e3e3cd213..4e34d8e4c1 100644
--- a/configs/p3541-0000_defconfig
+++ b/configs/p3541-0000_defconfig
@@ -15,14 +15,14 @@ CONFIG_CMD_DFU=y
 CONFIG_CMD_GPIO=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_MMC=y
-CONFIG_CMD_PCI=y
+#CONFIG_CMD_PCI=y
 CONFIG_CMD_SF=y
 CONFIG_CMD_SPI=y
-CONFIG_CMD_USB=y
-CONFIG_CMD_USB_MASS_STORAGE=y
-# CONFIG_CMD_SETEXPR is not set
+#CONFIG_CMD_USB=y
+#CONFIG_CMD_USB_MASS_STORAGE=y
+CONFIG_CMD_SETEXPR=y
 # CONFIG_CMD_NFS is not set
-CONFIG_CMD_EXT4_WRITE=y
+#CONFIG_CMD_EXT4_WRITE=y
 CONFIG_OF_LIVE=y
 CONFIG_DEFAULT_DEVICE_TREE="tegra210-p3541-0000"
 CONFIG_DFU_MMC=y
@@ -34,38 +34,45 @@ CONFIG_SPI_FLASH_MACRONIX=y
 CONFIG_SPI_FLASH_USE_4K_SECTORS=y
 CONFIG_SF_DEFAULT_MODE=0
 CONFIG_SF_DEFAULT_SPEED=24000000
-CONFIG_RTL8169=y
-CONFIG_PCI=y
-CONFIG_DM_PCI=y
-CONFIG_DM_PCI_COMPAT=y
-CONFIG_PCI_TEGRA=y
+#CONFIG_RTL8169=y
+#CONFIG_PCI=y
+#CONFIG_DM_PCI=y
+#CONFIG_DM_PCI_COMPAT=y
+#CONFIG_PCI_TEGRA=y
 CONFIG_SYS_NS16550=y
 CONFIG_TEGRA114_SPI=y
 CONFIG_TEGRA210_QSPI=y
-CONFIG_USB=y
-CONFIG_DM_USB=y
-CONFIG_USB_EHCI_HCD=y
-CONFIG_USB_EHCI_TEGRA=y
-CONFIG_USB_GADGET=y
-CONFIG_USB_GADGET_MANUFACTURER="NVIDIA"
-CONFIG_USB_GADGET_VENDOR_NUM=0x0955
-CONFIG_USB_GADGET_PRODUCT_NUM=0x701a
+#CONFIG_USB=y
+#CONFIG_DM_USB=y
+#CONFIG_USB_EHCI_HCD=y
+#CONFIG_USB_EHCI_TEGRA=y
+#CONFIG_USB_GADGET=y
+#CONFIG_USB_GADGET_MANUFACTURER="NVIDIA"
+#CONFIG_USB_GADGET_VENDOR_NUM=0x0955
+#CONFIG_USB_GADGET_PRODUCT_NUM=0x701a
 CONFIG_CI_UDC=y
-CONFIG_USB_GADGET_DOWNLOAD=y
-CONFIG_USB_HOST_ETHER=y
-CONFIG_USB_ETHER_ASIX=y
+#CONFIG_USB_GADGET_DOWNLOAD=y
+#CONFIG_USB_HOST_ETHER=y
+#CONFIG_USB_ETHER_ASIX=y
 # CONFIG_ENV_IS_IN_MMC is not set
-CONFIG_ENV_IS_IN_SPI_FLASH=y
-CONFIG_ENV_SIZE=0x2000
+CONFIG_ENV_IS_NOWHERE=y
+CONFIG_ENV_SIZE=0x4000
 CONFIG_ENV_SECT_SIZE=0x1000
 CONFIG_ENV_OFFSET=0xFFFD8000
-CONFIG_BOOTP_PREFER_SERVERIP=y
+#CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_POSITION_INDEPENDENT=y
 CONFIG_DISABLE_SDMMC1_EARLY=y
-CONFIG_NVME=y
-CONFIG_USB_XHCI_HCD=y
-CONFIG_USB_XHCI_TEGRA=y
-CONFIG_USB_STORAGE=y
+#CONFIG_NVME=y
+#CONFIG_USB_XHCI_HCD=y
+#CONFIG_USB_XHCI_TEGRA=y
+#CONFIG_USB_STORAGE=y
 CONFIG_DOS_PARTITION=y
 CONFIG_CMD_CACHE=y
 CONFIG_OF_LIBFDT_OVERLAY=y
+CONFIG_CMD_PART=y
+CONFIG_PARTITION_UUIDS=y
+CONFIG_CMD_IMPORTENV=y
+CONFIG_BOOTDELAY=0
+CONFIG_CMD_EXPORTENV=y
+CONFIG_CMD_FS_UUID=y
+CONFIG_AUTOBOOT=y
diff --git a/include/config_distro_bootcmd.h b/include/config_distro_bootcmd.h
index 599e2b7212..7b1f3b82f7 100644
--- a/include/config_distro_bootcmd.h
+++ b/include/config_distro_bootcmd.h
@@ -31,6 +31,8 @@
 			"run scan_dev_for_boot_part; " \
 		"fi\0"
 
+#define BOOT_TARGET_DEFAULTDEVPLIST    "C"
+
 #define BOOTENV_SHARED_BLKDEV(devtypel) \
 	#devtypel "_boot=" \
 	BOOTENV_SHARED_BLKDEV_BODY(devtypel)
@@ -434,6 +436,7 @@
 	"boot_script_dhcp=boot.scr.uimg\0" \
 	BOOTENV_BOOT_TARGETS \
 	\
+	"board_fdt_file=" __stringify(TEGRA_BOARD_FDT_FILE) "\0" \
 	"boot_syslinux_conf=extlinux/extlinux.conf\0" \
 	"boot_extlinux="                                                  \
 		"sysboot ${devtype} ${devnum}:${distro_bootpart} any "    \
@@ -449,7 +452,7 @@
 		"fi\0"                                                    \
 	\
 	"boot_a_script="                                                  \
-		"load ${devtype} ${devnum}:${distro_bootpart} "           \
+		"load ${resin_dev_type} ${resin_dev_index}:${resin_boot_part} "           \
 			"${scriptaddr} ${prefix}${script}; "              \
 		"source ${scriptaddr}\0"                                  \
 	\
@@ -476,7 +479,8 @@
 		"\0"                                                      \
 	\
 	"scan_dev_for_boot_part="                                         \
-		"part list ${devtype} ${devnum} -bootable devplist; "     \
+		"setenv devplist $resin_root_part; "                      \
+		"env exists devplist || part list ${resin_dev_type} ${resin_dev_index} -bootable devplist; "     \
 		"env exists devplist || setenv devplist 1; "              \
 		"for distro_bootpart in ${devplist}; do "                 \
 			"if fstype ${devtype} "                           \
@@ -493,8 +497,12 @@
 		BOOTENV_SET_NVME_NEED_INIT                                \
 		BOOTENV_SET_IDE_NEED_INIT                                 \
 		BOOTENV_SET_VIRTIO_NEED_INIT                              \
+		"setenv resin_kernel_load_addr ${loadaddr}; "             \
 		"for target in ${boot_targets}; do "                      \
-			"run bootcmd_${target}; "                         \
+                       "if env exist resin_set_kernel_root; then "       \
+                               "run resin_set_kernel_root; run set_os_cmdline; "  \
+                               "run bootcmd_${target}; "                 \
+                       "fi; " \
 		"done\0"
 
 #ifndef CONFIG_BOOTCOMMAND
diff --git a/include/configs/p3450-0000.h b/include/configs/p3450-0000.h
index de1410ade0..d01f2ddbdf 100644
--- a/include/configs/p3450-0000.h
+++ b/include/configs/p3450-0000.h
@@ -10,6 +10,18 @@
 
 #include "tegra210-common.h"
 
+#ifdef CONFIG_P3450_EMMC
+#define CONFIG_TEGRA_BOARD_STRING	"NVIDIA P3450-0002"
+#define BOOT_TARGET_DEVICES(func) \
+	func(MMC, mmc, 0) \
+	func(PXE, pxe, na) 
+
+#define CONFIG_SYS_MMC_ENV_DEV		0
+#define CONFIG_SYS_MMC_ENV_PART		2
+#define DEFAULT_MMC_DEV			"0"
+
+#else
+
 /* High-level configuration options */
 #define CONFIG_TEGRA_BOARD_STRING	"NVIDIA P3450-0000"
 
@@ -19,16 +31,16 @@
 /* Both SD and EMMC since this supports both SKU0 and SKU2 Nano */
 #define BOOT_TARGET_DEVICES(func) \
 	func(MMC, mmc, 1) \
-	func(MMC, mmc, 0) \
-	func(USB, usb, 0) \
-	func(NVME, nvme, 0) \
-	func(PXE, pxe, na) \
-	func(DHCP, dhcp, na)
+	func(PXE, pxe, na) 
 
 /* Environment at end of QSPI on the Nano devkit, in the VER partition */
 #define CONFIG_ENV_SPI_MAX_HZ		48000000
 #define CONFIG_ENV_SPI_MODE		SPI_MODE_0
 #define CONFIG_SPI_FLASH_SIZE		(4 << 20)
+#define DEFAULT_MMC_DEV                        "1"
+/* Crystal is 38.4MHz. clk_m runs at half that rate */
+#define COUNTER_FREQUENCY       19200000
+#endif
 
 #define CONFIG_PREBOOT
 
@@ -42,9 +54,6 @@
 #include "tegra-common-usb-gadget.h"
 #include "tegra-common-post.h"
 
-/* Crystal is 38.4MHz. clk_m runs at half that rate */
-#define COUNTER_FREQUENCY	19200000
-
 /*
  * Environment in eMMC only on the Nano commercial SKU,
  * at the end of 2nd "boot sector"
diff --git a/include/configs/p3541-0000.h b/include/configs/p3541-0000.h
index d42be12d7a..c0f8918f4b 100644
--- a/include/configs/p3541-0000.h
+++ b/include/configs/p3541-0000.h
@@ -19,9 +19,7 @@
 /* Nano2GB doesn't have eMMC or NVMe, just SD/USB/Net */
 #define BOOT_TARGET_DEVICES(func) \
 	func(MMC, mmc, 1) \
-	func(USB, usb, 0) \
-	func(PXE, pxe, na) \
-	func(DHCP, dhcp, na)
+	func(PXE, pxe, na)
 
 /* Environment at end of QSPI, in the VER partition */
 #define CONFIG_ENV_SPI_MAX_HZ		48000000
-- 
2.17.1

